File: app/src/commands/infrastructure/workshop.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { invoke } from "@tauri-apps/api/core";
import type {
  AddWorkshopRecruitOrderRequest,
  CancelWorkshopRecruitOrderRequest,
  Coord,
  GetWorkshopRecruitCatalogRequest,
  GetWorkshopRecruitCatalogResponse,
  WorkshopRecruitOrderId,
  WorkshopRecruitOrderRequest,
} from "@tsukilabs/nil-bindings";

export async function addWorkshopRecruitOrder(request: WorkshopRecruitOrderRequest) {
  const req: AddWorkshopRecruitOrderRequest = {
    world: NIL.world.getIdStrict(),
    request,
  };

  await invoke("add_workshop_recruit_order", { req });
}

export async function cancelWorkshopRecruitOrder(coord: Coord, id: WorkshopRecruitOrderId) {
  const req: CancelWorkshopRecruitOrderRequest = {
    world: NIL.world.getIdStrict(),
    coord,
    id,
  };

  await invoke("cancel_workshop_recruit_order", { req });
}

export async function getWorkshopRecruitCatalog(coord: Coord) {
  const req: GetWorkshopRecruitCatalogRequest = {
    world: NIL.world.getIdStrict(),
    coord,
  };

  return invoke<GetWorkshopRecruitCatalogResponse>("get_workshop_recruit_catalog", { req });
}

File: app/src-tauri/src/command/infrastructure/workshop.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::error::Result;
use crate::manager::ManagerExt;
use nil_core::infrastructure::building::workshop::recruit_catalog::WorkshopRecruitCatalog;
use nil_payload::infrastructure::workshop::*;
use tauri::AppHandle;

#[tauri::command]
pub async fn add_workshop_recruit_order(
  app: AppHandle,
  req: AddWorkshopRecruitOrderRequest,
) -> Result<()> {
  app
    .client(async |cl| cl.add_workshop_recruit_order(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn cancel_workshop_recruit_order(
  app: AppHandle,
  req: CancelWorkshopRecruitOrderRequest,
) -> Result<()> {
  app
    .client(async |cl| cl.cancel_workshop_recruit_order(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_workshop_recruit_catalog(
  app: AppHandle,
  req: GetWorkshopRecruitCatalogRequest,
) -> Result<WorkshopRecruitCatalog> {
  app
    .client(async |cl| cl.get_workshop_recruit_catalog(req).await)
    .await
    .map_err(Into::into)
}
